### PR TITLE
Used hardware PRNG for dropout

### PR DESCRIPTION
**Pull request by amahmudTT**
Implemented the dropout. 

Used 31 bits from PRNG generator for perf.
Used 32 bit float for scale factor, in BUDA it is using 16 bits.

The seed initialization is very slow and requires 500 NOPS at least. 
We could pull some code out of the dropout function and put in the initialization for performance but I tried to follow existing code. 

Post commit tests pass (it was not being used by anything else) , I will run them again though when I make the pr to tt-metal.
https://github.com/tenstorrent/tt-metal/actions/runs/10087888953

### CODE REVIEW

**Comment by amahmudTT** [comment]
It seems the PRNG is very crude (as long as my code is correct). 

I created a test that uses a vector of size 131072 and applied differing probabilities for dropout. 
The PRNG showed great dependence on the seed value and here are some example outputs.

```
                   Verif | INFO     | Created a constant vector of size 131072 with value 9, bf16 = 1091584272
                   Test | INFO     | dropout probability=0, dropout_rate=0               
                   Test | INFO     | dropout probability=0.1, dropout_rate=0.11258316
                   Test | ERROR    | Dropout rate & probability mismatch probability=0.2, dropout_rate=0.1400299
                   Test | INFO     | dropout probability=0.3, dropout_rate=0.31822968
                   Test | ERROR    | Dropout rate & probability mismatch probability=0.4, dropout_rate=0.4645157
                   Test | INFO     | dropout probability=0.5, dropout_rate=0.5048866
                   Test | ERROR    | Dropout rate & probability mismatch probability=0.6, dropout_rate=0.520134
                   Test | ERROR    | Dropout rate & probability mismatch probability=0.7, dropout_rate=0.573844
                   Test | ERROR    | Dropout rate & probability mismatch probability=0.8, dropout_rate=0.8557396
                   Test | INFO     | dropout probability=0.9, dropout_rate=0.9181709
                   Test | INFO     | dropout probability=1, dropout_rate=1
```

Half the time the dropout rate deviated quite a bit from probability.  It seems it achieves good results when the rates are 
0.25, 0.5  (0&1 are trivial). Which signifies the algorithm was tuned to distribute the numbers evenly in halves/quarters but not overall.

**Comment by ttmtrajkovic** [comment]
This is great Anil.
Every SFPU has its own PRNG and given that there are 32 SFPUs, every PRNG will have `total_number / 32` samples. every PRNG should be producing uniform distribution, as long as there's sufficient number of samples.
also, if you are using 31 bits from PRNG as your random sample, is your probability also scaled to 31 bits?

let's talk about seed init on Monday.

File: common/inc/sfpu/ckernel_sfpu_dropout.h
Comment (ttmtrajkovic) [inline common/inc/sfpu/ckernel_sfpu_dropout.h:58]: You shouldnt need to enable CC explicitly as by default CC is always enabled.
you just have to worry about reverting it to the correct state if CC is modified using your insstruction

**Comment by amahmudTT** [comment]
> This is great Anil. Every SFPU has its own PRNG and given that there are 32 SFPUs, every PRNG will have `total_number / 32` samples. every PRNG should be producing uniform distribution, as long as there's sufficient number of samples. also, if you are using 31 bits from PRNG as your random sample, is your probability also scaled to 31 bits?
> 
> let's talk about seed init on Monday.

Yes I did scale it but I assume its the responsibility of the op to make sure the input is that way.

**Comment by amahmudTT** [comment]
I actually had a bug in my test kernel , I tried to follow other kernels and initialized the PRNG before every tile. So effectively my results reflected the RNG for only one tile and got inaccurate results. After initializing only once before all the tiles I get very good results. 
```
                   Test | INFO     | dropout probability=0, dropout_rate=0
                   Test | INFO     | dropout probability=0.1, dropout_rate=0.10077286
                   Test | INFO     | dropout probability=0.2, dropout_rate=0.20359802
                   Test | INFO     | dropout probability=0.3, dropout_rate=0.29244232
                   Test | INFO     | dropout probability=0.4, dropout_rate=0.42268372
                   Test | INFO     | dropout probability=0.5, dropout_rate=0.5007248
                   Test | INFO     | dropout probability=0.6, dropout_rate=0.6022377
                   Test | INFO     | dropout probability=0.7, dropout_rate=0.7030449
                   Test | INFO     | dropout probability=0.8, dropout_rate=0.80612946
                   Test | INFO     | dropout probability=0.9, dropout_rate=0.8924713
                   Test | INFO     | dropout probability=1, dropout_rate=1
```

**Comment by ttmtrajkovic** [comment]
> I actually had a bug in my test kernel , I tried to follow other kernels and initialized the PRNG before every tile. So effectively my results reflected the RNG for only one tile and got inaccurate results. After initializing only once before all the tiles I get very good results.
> 
> ```
>                    Test | INFO     | dropout probability=0, dropout_rate=0
>                    Test | INFO     | dropout probability=0.1, dropout_rate=0.10077286
>                    Test | INFO     | dropout probability=0.2, dropout_rate=0.20359802
>                    Test | INFO     | dropout probability=0.3, dropout_rate=0.29244232
>                    Test | INFO     | dropout probability=0.4, dropout_rate=0.42268372
>                    Test | INFO     | dropout probability=0.5, dropout_rate=0.5007248
>                    Test | INFO     | dropout probability=0.6, dropout_rate=0.6022377
>                    Test | INFO     | dropout probability=0.7, dropout_rate=0.7030449
>                    Test | INFO     | dropout probability=0.8, dropout_rate=0.80612946
>                    Test | INFO     | dropout probability=0.9, dropout_rate=0.8924713
>                    Test | INFO     | dropout probability=1, dropout_rate=1
> ```

This looks great. The distribution should be even better for larger data sets

**Comment by amahmudTT** [comment]
Post commits passes with this branch https://github.com/tenstorrent/tt-metal/actions/runs/10200615586
I think I can merge this one, once approved. I separated the call to PRNG initialization at seems multiple functions will need to call it.

**Review by rdjogoTT — APPROVED** [review]

**Comment by corsix** [comment]
> This is great Anil. Every SFPU has its own PRNG and given that there are 32 SFPUs, every PRNG will have `total_number / 32` samples. every PRNG should be producing uniform distribution

How does the single seed get distributed to the 32 SFPUs? In my testing, I'm seeing all 32 vector lanes generate the same random number, which is not great, and implies I'm missing a step to cause the lanes to diverge.

**Comment by amahmudTT** [comment]
> > This is great Anil. Every SFPU has its own PRNG and given that there are 32 SFPUs, every PRNG will have `total_number / 32` samples. every PRNG should be producing uniform distribution
> 
> How does the single seed get distributed to the 32 SFPUs? In my testing, I'm seeing all 32 vector lanes generate the same random number, which is not great, and implies I'm missing a step to cause the lanes to diverge.

Are you trying to generate random numbers in a separate kernel ? Could you try initializing the seed and see if you get the same result ? 

```init_prng_seed```

As for the dropout  OP in python is implemented in such a way that the seed is reinitialized for every tile. Thus all the tiles will have similar pattern and this is going to be addressed in the future.

**Comment by amahmudTT** [comment]
> My exact setup is somewhat bespoke, so I've tried to reproduce what I'm seeing using a minimally modified tt_metal 

Hello Peter, 

Sorry for the delay in answering and thank you very much for pointing this issue.  We have created an issue https://github.com/tenstorrent/tt-metal/issues/13904 to to track this problem and will try to fix it. Thanks again.

**Pull request merged**
